apiserver/facades/controller/crossmodelsecrets/crossmodelsecrets.go

```diff
@@ -0,0 +1,179 @@
+type secretStateGetter func(modelUUID string) (SecretsState, SecretsConsumer, func() bool, error)
+
+// CrossModelSecretsAPI provides access to the CrossModelSecrets API facade.
+type CrossModelSecretsAPI struct {
+	resources facade.Resources
+
+	ctx       stdcontext.Context
+	mu        sync.Mutex
+	modelUUID string
+
+	secretsStateGetter  secretStateGetter
+	backendConfigGetter backendConfigGetter
+	crossModelState     CrossModelState
+}
+
+// NewCrossModelSecretsAPI returns a new server-side CrossModelSecretsAPI facade.
+func NewCrossModelSecretsAPI(
+	resources facade.Resources,
+	modelUUID string,
+	secretsStateGetter secretStateGetter,
+	backendConfigGetter backendConfigGetter,
+	crossModelState CrossModelState,
+) (*CrossModelSecretsAPI, error) {
+	return &CrossModelSecretsAPI{
+		ctx:                 stdcontext.Background(),
+		resources:           resources,
+		modelUUID:           modelUUID,
+		secretsStateGetter:  secretsStateGetter,
+		backendConfigGetter: backendConfigGetter,
+		crossModelState:     crossModelState,
+	}, nil
+}
+
+var logger = loggo.GetLogger("juju.apiserver.crossmodelsecrets")
+
+func (s *CrossModelSecretsAPI) checkMacaroonsForConsumer(consumerTag names.Tag, secretURI string, mac macaroon.Slice, version bakery.Version) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// TODO(cmr secrets)
+	return nil
+}
+
+// GetSecretContentInfo returns the secret values for the specified secrets.
+func (s *CrossModelSecretsAPI) GetSecretContentInfo(args params.GetRemoteSecretContentArgs) (params.SecretContentResults, error) {
+	result := params.SecretContentResults{
+		Results: make([]params.SecretContentResult, len(args.Args)),
+	}
+	for i, arg := range args.Args {
+		content, backend, err := s.getSecretContent(arg)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		contentParams := params.SecretContentParams{}
+		if content.ValueRef != nil {
+			contentParams.ValueRef = &params.SecretValueRef{
+				BackendID:  content.ValueRef.BackendID,
+				RevisionID: content.ValueRef.RevisionID,
+			}
+		}
+		if content.SecretValue != nil {
+			contentParams.Data = content.SecretValue.EncodedValues()
+		}
+		result.Results[i].Content = contentParams
+		result.Results[i].BackendConfig = backend
+	}
+	return result, nil
+}
+
+func (s *CrossModelSecretsAPI) getSecretContent(arg params.GetRemoteSecretContentArg) (*secrets.ContentParams, *params.SecretBackendConfigResult, error) {
+	if arg.URI == "" {
+		return nil, nil, errors.NewNotValid(nil, "empty uri")
+	}
+	uri, err := coresecrets.ParseURI(arg.URI)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	if uri.SourceUUID == "" {
+		return nil, nil, errors.NotValidf("secret URI with empty source UUID")
+	}
+
+	app, err := s.crossModelState.GetRemoteEntity(arg.ApplicationToken)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	consumer := names.NewUnitTag(fmt.Sprintf("%s/%d", app.Id(), arg.UnitId))
```

Comment:
Might be better to define GetRemoteSecretContentArg.UnitId as a pointer

Reply:
We always need the unit id to be set since it is a unit agent making the api call.